data_safe_haven/utility/logger.py

```diff
@@ -0,0 +1,234 @@
+
+    def format_msg(self, message: str, level: int = logging.INFO) -> str:
+        """Format a message using rich handler"""
+        for handler in self.logger.handlers:
+            if isinstance(handler, RichHandler):
+                fn, lno, func, sinfo = self.logger.findCaller(False, 1)
+                return handler.format(
+                    self.logger.makeRecord(
+                        name=self.logger.name,
+                        level=level,
+                        fn=fn,
+                        lno=lno,
+                        msg=message,
+                        args={},
+                        exc_info=None,
+                        func=func,
+                        sinfo=sinfo,
+                    )
+                )
+        return message
+
+    def style(self, message: str) -> str:
+        """Apply logging style to a string"""
+        markup = self.format_msg(message)
+        return RichStringAdaptor(coloured=True).to_string(markup)
+
+    # Pass log levels through to the logger
+    def critical(self, message: str) -> None:
+        return self.logger.critical(message)
+
+    def error(self, message: str) -> None:
+        return self.logger.error(message)
+
+    def warning(self, message: str) -> None:
+        return self.logger.warning(message)
+
+    def info(self, message: str) -> None:
+        return self.logger.info(message)
+
+    def debug(self, message: str) -> None:
+        return self.logger.debug(message)
+
+    # Loggable wrappers for confirm/ask/choice
+    def confirm(self, message: str, default_to_yes: bool = True) -> bool:
+        formatted = self.format_msg(message, logging.INFO)
+        return Confirm.ask(formatted, default=default_to_yes)
+
+    def ask(self, message: str, default: Optional[str] = None) -> str:
+        formatted = self.format_msg(message, logging.INFO)
+        if default:
+            return Prompt.ask(formatted, default=default)
+        return Prompt.ask(formatted)
+
+    def choose(
+        self,
+        message: str,
+        choices: Optional[List[str]] = None,
+        default: Optional[str] = None,
+    ) -> str:
+        formatted = self.format_msg(message, logging.INFO)
+        if default:
+            return Prompt.ask(formatted, choices=choices, default=default)
+        return Prompt.ask(formatted, choices=choices)
+
+    # Apply a level to non-leveled messages
+    def parse(self, message: str) -> None:
+        tokens = message.split(":")
+        level, remainder = tokens[0], ":".join(tokens[1:]).strip()
+        if level == "CRITICAL":
+            return self.critical(remainder)
+        elif level == "ERROR":
+            return self.error(remainder)
+        elif level == "WARNING":
+            return self.warning(remainder)
+        elif level == "INFO":
+            return self.info(remainder)
+        elif level == "DEBUG":
+            return self.debug(remainder)
+        else:
+            return self.info(message.strip())
```

Comment:
What is this applied to? I can see cases where it is run on the output of Azure CLI/API. Is there a more specific name for the method?

Reply:
I thought `logger.parse` -> parses log messages was reasonably clear, but I may have been too terse here! What would you call a function that is processing plain-text messages and emitting an appropriate log message?

Reply:
Oh yes, I didn't quite mean that.

The method assumes things about the form of the input, that it will be `<level>:<message>` and that level is one of a set of values (otherwise we assume the level is info).

It felt like maybe this method was to parse some particular log output rather than log messages from _any_ program.

Reply:
It's to parse log messages that we produce from systems that can't directly talk to our logging code (eg. when we run external commands on a VM and get stdout back but we want to present it as log messages).

Reply:
Do we know that those will all give log messages that play nicely with this function?

I can see the need. Maybe just adding a docstring to say what this function is for is the best thing to do for now.

Reply:
To be clear: this is to parse messages from **non-Python** code in the DSH codebase.